src/main/java/org/opensearch/ad/cluster/ADDataMigrator.java

```diff
@@ -0,0 +1,292 @@
+import org.opensearch.common.xcontent.NamedXContentRegistry;
+import org.opensearch.common.xcontent.XContentFactory;
+import org.opensearch.common.xcontent.XContentParser;
+import org.opensearch.index.IndexNotFoundException;
+import org.opensearch.index.query.BoolQueryBuilder;
+import org.opensearch.index.query.MatchAllQueryBuilder;
+import org.opensearch.index.query.TermQueryBuilder;
+import org.opensearch.index.query.TermsQueryBuilder;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+/**
+ * Migrate AD data to support backward compatibility.
+ * Currently we need to migrate:
+ *   1. Detector internal state (used to track realtime job error) to realtime data.
+ */
+public class ADDataMigrator {
+    private final Logger logger = LogManager.getLogger(this.getClass());
+    private final Client client;
+    private final ClusterService clusterService;
+    private final NamedXContentRegistry xContentRegistry;
+    private final AnomalyDetectionIndices detectionIndices;
+    private final AtomicBoolean dataMigrated;
+
+    public ADDataMigrator(
+        Client client,
+        ClusterService clusterService,
+        NamedXContentRegistry xContentRegistry,
+        AnomalyDetectionIndices detectionIndices
+    ) {
+        this.client = client;
+        this.clusterService = clusterService;
+        this.xContentRegistry = xContentRegistry;
+        this.detectionIndices = detectionIndices;
+        this.dataMigrated = new AtomicBoolean(false);
+    }
+
+    /**
+     * Migrate AD data. Currently only need to migrate detector internal state {@link DetectorInternalState}
+     */
+    public void migrateData() {
+        if (!dataMigrated.getAndSet(true)) {
+            logger.info("Start migrating AD data");
+
+            if (!detectionIndices.doesAnomalyDetectorJobIndexExist()) {
+                logger.info("AD job index doesn't exist, no need to migrate");
+                return;
+            }
+
+            if (detectionIndices.doesDetectorStateIndexExist()) {
+                migrateDetectorInternalStateToRealtimeTask();
+            } else {
+                // If detection index doesn't exist, create index and backfill realtime task.
+                detectionIndices.initDetectionStateIndex(ActionListener.wrap(r -> {
+                    if (r.isAcknowledged()) {
+                        logger.info("Created {} with mappings.", CommonName.DETECTION_STATE_INDEX);
+                        migrateDetectorInternalStateToRealtimeTask();
+                    } else {
+                        String error = "Create index " + CommonName.DETECTION_STATE_INDEX + " with mappings not acknowledged";
+                        logger.warn(error);
+                    }
+                }, e -> {
+                    if (ExceptionsHelper.unwrapCause(e) instanceof ResourceAlreadyExistsException) {
+                        migrateDetectorInternalStateToRealtimeTask();
```

Comment:
How would this exception occur? Is this to handle some race condition where it gets created some time after the `if` check on line 102?

Reply:
Yes, while we migrate data, it's possible that user run some historical detector and it will create detection state index. Then we will see such exception.